test/alternator: functional tests for Alternator multi-item transaction feature

nyh opened the pull request:
This patch adds extensive functional tests for the DynamoDB multi-item transactions feature - the TransactWriteItems and TransactGetItems requests. We add 43 test functions, spanning more than 1000 lines of code, covering the different parameters and corner cases of these requests.
    
Because we don't support the transaction feature in Alternator yet (this is issue #5064), all of these tests fail on Alternator but all of them were tested to pass on DynamoDB. So all new tests are marked "xfail".
    
These tests will be handy for whoever will implement this feature as an acceptance test, and can also be useful for whoever will just want to understand this feature better - the tests are short and simple and heavily commented. 
    
Note that these tests only check the correct functionality of individual calls of these requests - these tests cannot and do not check the consistency or isolation guarantees of concurrent invocations of several requests. Such tests would require a different test framework, such as the one requested in issue #6350, and are therefore not part of this patch.
    
Note that this patch includes ONLY tests, and does not mean that an implementation of the feature will soon follow. In fact, nobody is currently working on implementing this feature.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14387/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14387/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1101/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/504/)
#### Build Details:
- Duration: 5 hr 47 min
- Builder: spider4.cloudius-systems.com

nyh commented:
Copyedited and pushed a new version.
The newest version just has (a lot...) of typo fixes in comments, no functional changes to the tests.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14435/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14435/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 12 hr
- Builder: i-0bab6e7b68c7169c8 (m5ad.8xlarge)

nyh commented:
Please review and consider merging. This is a "harmless" set of xfailing tests (passing on DynamoDB, failing on Alternator) for the "transactions" feature that we don't have yet. I think that it can really help anyone who wants to understand this feature better (as it explores the various corner cases of the DynamoDB API) and certainly for anyone in the future who might want to implement it.

swasik commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+                'ExpressionAttributeNames': {'#xx': 'x'}
+            }}])
+
+# Test the ability to read multiple items in one transaction. We can actually
```
You have the same for the write somewhere in the middle of the file. Maybe we should keep it together for easier reading.

nyh replied:
Yes, but on the other hand I decided to bunch all the read tests together and the all the write tests together, so this is why this test is together with the other write tests. So I'm not sure which of the two classifications is more helpful. Maybe you're right, I'll think about it again.

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+# one item.
+@pytest.mark.xfail(reason="#5064 - transactions not yet supported")
+def test_transact_write_items_multi_table_true(test_table_s, test_table):
+    p1 = random_string()
```
Oh, the "_s" just means the key is a single string. I also have "_sn" (the key is a string and a number) and a bunch of others, defined in conftest.py. The idea is that if we have a handful of tables shared by many tests, each individual test doesn't even need to create a table - which is the slowest part of each test. Thanks to this approach, many of these tests take milliseconds (!) to run.

Anyway, "test_table", without the prefix, was just the very first table I created in conftest.py. Actually, I do have another name, `test_table_ss` defined in conftest.py. Maybe I should indeed use that name for clarity, not test_table.

(by the way, it was also a mistake to name these fixtures test_something - it makes them sound like they are tests! I invented this name before I became a pytest expert... But it's not that terrible so I never went back to rename them in 1000 tests).

nyh replied:
I changed the usage of "test_table" in this test to "test_table_ss", which I agree is more self-descriptive.
There are 142 existing tests using the name "test_table", I won't be changing all of them in this patch (obviously).  We can consider doing this later, I'll open an issue about it.

nyh replied:
https://github.com/scylladb/scylladb/issues/22494

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+            'TableName': test_table_s.name,
+            'Item': item,
+        }} for item in items])
+    # verify that all 100 items went it
```
Right.

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+    assert reasons[0]['Code'] == 'None'
+    assert reasons[1]['Code'] == 'None'
+    assert reasons[2]['Code'] == 'ConditionalCheckFailed'
+    assert reasons[3]['Code'] == 'ConditionalCheckFailed'
```
Yes, all the tests passed on DynamoDB (in fact, that's the only thing I could run them on, since we don't have this feature on Alternator). 

And yes, it seems that all conditions are checked before making a decision, there is no "short circuiting" like in C's "&&". However, as I uncovered in other tests, if the condition has a **mistake** it's not just false (syntax error, wrong type, etc.) it is possible that only the first error is recognized, and in some cases we don't even get a list of errors and just get a single failure on the entire transaction. I'm not sure it will be very important to replicate 100% each and every one of these details in our own implementation, but in general it's a good idea to understand the rules before breaking them. If after careful consideration we decide that in our implementation we will deviate from one DynamoDB's behaviors, I can change the test to accept both behaviors, or remove the test, and so on. When I (rarely) do that, I usually add a long comment in the test, so the test suite becomes an interesting read for people wondering how DynamoDB, Alternator, or both, behave - and why.

nyh replied:
I'll add some comments about this to the "book" in the test ;-)

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+                'Item': {'p': 7}
+            }}])
+    reasons = e.value.response['CancellationReasons']
+    assert reasons[0]['Code'] == 'ValidationError'
```
I was lazy to check - I didn't think it was interesting because I checked the message in other tests that covered more than this test does, and the message wasn't it's focus.
But I can add the message check here too. I'll do it.

nyh replied:
Added the check.

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+                'ExpressionAttributeNames': {'#xyz': 7}
+            }}])
+
+# TODO: check size limits on total write size (we already checked above
```
Not really. It's slower and more expensive than other tests because if X is the limit, we need to also check that a request with X-1 bytes does work. But this shouldn't be prohibitive (it will still cost less than a cent to run).
I'll do it, just to get rid of the last TODO :-) (as you can see, this test file has 1000 lines of code - it started with dozens of TODO lines and then I wrote the tests one by one until this was just the last one left).

nyh replied:
By the way if you're curious about the "more expensive" part, well, I calculated writing 3MB (to verify that 3MB is allowed) in a transaction costs 6,000 WCU (that's 2 WCU per 1 KB), which at $0.625 per million WCU (the new lowered prices) costs 0.375 cents. 
I think we'll survive that cost :-) Especially that I'll probably only ever run this test 10 or so times.

nyh commented on test/alternator/test_transact.py:
```diff
@@ -0,0 +1,974 @@
+# Single Update action without a condition
+# Unlike the previous test_transact_write_items_single_* tests which used
+# trivial conditions without ExpressionAttributeNames and
+# ExpressionAttributeValues, the following tests do use those (this test
```
I'll do that.
Note that this point (the last three lines of this comment) isn't really an important "section" of the test. It's just a comment that although in the previous tests we simplified the API usage by not using ExpressionAttributeNames/Values, we also need to start testing those so we have them in the following tests.

nyh commented:
Pushed a new version addressing all review comments (see my individual responses), and adding 4 more tests - 3 covering the transaction size limit that previously was a TODO (now there are no more TODOs), and a fourth testing that a Put action replaces an item (that's "obvious" seeing that's how PutItem behaves, but in Scylla it is far from obvious because that's not how a CQL "Insert" behaves).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14740/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14740/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 6 hr 38 min
- Builder: spider3.cloudius-systems.com

nyh commented:
@scylladb/scylla-maint please consider merging.
